### PR TITLE
High precision time units

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
This patch introduces new time units of microsecond and
nanosecond precision forming a new 'time-units-small' category.

Also found and fixed several problems, related to time parameters
parsing:

* Obscure "integer overflow" fatal messages. For example, passing
  "0.0001 second" caused this message. After fixing, Squid reports
  that the value "is too small to be used in this context".
    
* Ignoring possible zero-rounded values after parsing. For example, if 
  a second-precision parameter was configured with 0.1 second, it
  became zero after rounding, which is unexpected. It is treated
  as a fatal error now.
 
* Inconsistent parameter overflow type. For example, parameters
  with millisecond and second precision reported that 'time_msec_t'
  overflowed. Now we introduce an absolute time maximum allowed,
  equal to the maximum of chrono::nanoseconds type which is about
  293 years. This absolute maximum allows to keep the time parsing
  code simple and at the same time should satisfy any reasonable
  configuration need. Note that this solution treats existing
  configurations with unreasonably huge time values > 293 years
  as fatal errors, such configurations should be fixed accordingly.

* Time overflows for icap_service_failure_limit parameter were not
  checked at all. This is probably a result of code duplication.
  By fixing the latter problem, the former one was resolved
  automatically.

* Unclear fatal message if a time parameter lacked time unit. Now
  Squid reports about "missing time unit".

* Improved error reporting when an inapplicable time unit was used, for
  example a 'millisecond' instead of a 'second'. For the majority of
  time parameters, it reported only a common "FATAL: Bungled..."
  message. For url_rewrite_timeout parameter, it reported an irrelevant
  "unsupported option ..." message (since it began to treat the faulty
  time unit as the next option). Now in both cases it reports about the
  underlying time unit problem.

While fixing these bugs I had to refactor and improve time parsing
functions, using safer std::chrono types instead of raw integer types.

### CODE REVIEW

**Comment by eduard-bagdasaryan** [comment]
The improvements and fixes, introduced in this PR were made while working on a Factory [project](https://github.com/measurement-factory/squid/pull/43), introducing a new 'paranoid_hit_validation' directive. So, in order to simplify/facilitate PR review, I decided to split the initial patch into two(relatively independent) parts. I'll be able to post the new feature itself as a separate PR once these (preliminary) changes are approved/merged into master.

**Comment by yadij** [comment]
This would be a good time to add a test-suite/squidconf/timeunits file with config lines that demonstrate the various ways we intend time units to be configured. If you can come up with variations that exercise any special edge cases too that would be great.

File: src/cache_cf.cc [hunk 1]
```diff
@@ -4873,11 +4939,43 @@ static void free_ftp_epsv(acl_access **ftp_epsv)
...
+    std::chrono::nanoseconds parsedUnitDuration;
+
+    if (parsedValue) {
+        const auto unitToken = parsedValue ? ConfigParser::PeekAtToken() : nullptr;
+        if (parseTimeUnit<Seconds>(unitToken, parsedUnitDuration))
+            (void)ConfigParser::NextToken();
+        else {
+            const auto defaultParsed = parseTimeUnit<Seconds>(T_SECOND_STR, parsedUnitDuration);
+            assert(defaultParsed);
+            debugs(3, DBG_PARSE_NOTE(DBG_IMPORTANT), "WARNING: " << ConfigPosition << "missing time unit, using deprecated default '" << T_SECOND_STR << "'");
```
Comment (yadij) [inline src/cache_cf.cc]: This message does not make any sense. It is only reachable past the assert when parseTimeUnit was *successful* parsing the unit.

File: src/cache_cf.cc [hunk 2]
```diff
@@ -4873,11 +4939,43 @@ static void free_ftp_epsv(acl_access **ftp_epsv)
...
+        self_destruct();
+
+    using Seconds = std::chrono::seconds;
+
+    const auto parsedValue = xatof(valueToken);
+
+    std::chrono::nanoseconds parsedUnitDuration;
+
+    if (parsedValue) {
+        const auto unitToken = parsedValue ? ConfigParser::PeekAtToken() : nullptr;
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/cache_cf.cc]: Done.

File: src/cache_cf.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/cache_cf.cc]: Since Squid time parameters (including url_rewrite_timeout parameter) do not allow default time units, this should be a fatal error. However, I decided to leave the old behavior (the default 'second' time unit for url_rewrite_timeout) as is but add a deprecation warning, so that it would be easier to fix this in the future.

File: src/cf.data.pre [hunk 3]
```diff
@@ -55,6 +55,24 @@ COMMENT_START
 		MB - Megabyte
 		GB - Gigabyte
 
+  Values with time units
+
+	Time-related directives marked with either "time-units" or
+	"time-units-small" accept a time unit. The supported time units are:
+
+		nanosecond - one billionth (10^-9) of a second (time-units-small only)
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/cf.data.pre]: Agreed and removed this obvious information except for the month (which probably could be confused between 30/31 days) and the suggested year.

File: src/cache_cf.cc [hunk 4]
```diff
@@ -536,9 +546,17 @@ parseOneConfigFile(const char *file_name, unsigned int depth)
...
+            } else {
+                try {
+                    if (!parse_line(tmp_line)) {
+                        debugs(3, DBG_CRITICAL, ConfigPosition << "unrecognized: '" << tmp_line << "'");
+                        ++err_count;
+                    }
+                } catch (...) {
+                    // fatal for now
+                    debugs(3, DBG_CRITICAL, "configuration error: " << CurrentException);
+                    self_destruct();
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/cache_cf.cc]: > Firstly it reverts a large amount of work done to make the -k parse feature maximize the amount of errors discovered.

I disagree: the old and the new code should behave similarly. The new try/catch near parse_line() results in catching and dying early within parseOneConfigFile(), while the old code would propagate the same exception up to parseConfigFile() and dying either. The 'trivial' errors reporting (which happen when parse_line() returns false because e.g., a parameter name is unknown), did not change: these errors were non-fatal and are still non-fatal. The new code should work just like the old code and report about all such cases. Basically, this PR only substitutes some self_destruct() with throw(), and in both cases '-k parse' aborts on the first error.

File: src/cache_cf.cc [hunk 5]
```diff
@@ -152,6 +155,8 @@ static const char *const B_GBYTES_STR = "GB";
 
 static const char *const list_sep = ", \t\n\r";
 
+static const double HoursPerYear = 24*365.2522;
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/cache_cf.cc]: I mean why we need to calculate from Hours. Theres no proper year in std::chrono ?

[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/cache_cf.cc]: There is a suitable std::chrono::years type, but it is [available](https://en.cppreference.com/w/cpp/chrono/duration) only _since C++20._

[the remark this answers is not in the document]
Comment (yadij) [reply src/cache_cf.cc]: then that is what the comment needs to be talking about. eg
```suggestion
// std::chrono::years requires C++20. Do our own rough calculation for now.
static const double HoursPerYear = 24*365.2522;
```

[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/cache_cf.cc]: Done at 88e60d7.

**Comment by eduard-bagdasaryan** [comment]
@yadij , have I addressed your requests? Please re-review.

File: src/cache_cf.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/cache_cf.cc]: This is the last point of disagreement here. We can resolve that with a file in test-suite/squidconf/ containing example config lines for all the directives affected by this change. It needs to test the expected and edge cases for the *old* configurable values so we can prove this change only adds new possibilities.

If in doing so you find some regressions from WARNING to ERROR those will need either fixing or documenting - decide on a case-by-case basis.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/cache_cf.cc]: > Amos: ... containing example config lines for all the directives affected by this change ...

If it is too difficult to collect _all_ of the directives that use modified time parsing routines, then adding at least one directive for each WARNING and non-fatal ERROR case (old and/or new) ought to be enough IMO. Please keep in mind that `test-suite/squidconf/` does not support test cases that are expected to fail so do not be tempted to add cases leading to Squid death.

If you add test cases that are expected to produce WARNINGs and ERRORs, please add squid.conf comments marking those test cases as such. Squid does not support proper testing of such cases (yet), but we should at least mark them when we add them.

[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/cache_cf.cc]: Done at 9455151.

> at least one directive for each WARNING and non-fatal ERROR case 

It is not possible to test related 'ERROR' cases, because all of them lead to Squid termination.
Both WARNING messages are displayed only when running '-k parse'.

**Comment by eduard-bagdasaryan** [comment]
Please mark as 'waiting-for-reviewer'.

**Comment by rousskov** [comment]
@yadij, I have not looked at the latest changes myself, but it looks like Eduard has tried to address your concerns (adding positive test cases), and this PR is on your side.

**Comment by yadij** [comment]
This is cleared for merge under the 10+ day criteria.

**Comment by yadij** [comment]
Arg, github buttons.
